File: HW2/Task 3&4/task4_python/Task4Merchant.py
from Pos import Pos

class Merchant():
    def __init__(self):
        self.elixirPrice = 1
        self.shieldPrice = 2
        self.pos = Pos()

    def talk(self, text):
        print("Merchant$: "+ text)

    def actionOnSoldier(self, soldier):
        fightEnabled = True
        while(fightEnabled):
            self.talk("Do you want to buy something? (1. Elixir, 2. Shield, 3.Leave.) Input: ")
            x = input()
            if x == '1':
                if(soldier.getCoins()>0):
                    print("=>You buy an elixir")
                    soldier.addElixiers()
                    soldier.useCoins(1)
                    fightEnabled = False
                else:
                    self.talk("You don't have enough coins.")
                    fightEnabled = False
            elif x == '2' :
                if(soldier.getCoins()>1):
                    print("=>You buy a shield")
                    soldier.useCoins(2)
                    soldier.addDefence()
                    fightEnabled = False
                else:
                    self.talk("You don't have enough coins.")
                    fightEnabled = False
            elif x == '3' :
                print("=> You leave.")
                fightEnabled = False
            else:
                print("=> Illegal choice!")
        
    def getPos(self):
        return self.pos

    def setPos(self, row, column):
        self.pos.setPos(row,column)
 
    def displaySymbol(self):
        print("$",end="")

File: HW2/Task 3&4/Java/Task4Merchant.java
import java.util.Scanner;

public class Merchant {
  private int elixirPrice;
  private int shieldPrice;
  private Pos pos;

  public Merchant() {
    // TODO: Initialization.
    this.elixirPrice = 1;
    this.shieldPrice = 2;
    this.pos = new Pos();
  }

  public void actionOnSoldier(Task4Soldier soldier) {
    this.talk("Do you want to buy something? (1. Elixir, 2. Shield, 3. Leave.) Input: ");

    // TODO: Main logic.
    boolean fightEnabled = true;
    while(fightEnabled)
    {
        this.talk("Do you want to buy something? (1. Elixir, 2.Shield, 3. Leave.) Input:%n ");
        Scanner sc = new Scanner(System.in);
        String choice = sc.nextLine();
        
        if(choice.equalsIgnoreCase("1"))
        {
            if(soldier.getCoins()>0){
                System.out.print("=>You buy an elixir.%n");
                soldier.addElixir();
                soldier.useCoins(1);
                fightEnabled = false;
            }else{
                this.talk("You don't have enough coins.%n");
                fightEnabled = false;
            }
        }else if(choice.equalsIgnoreCase("2"))
        {
            if(soldier.getCoins()>1){
                System.out.print("=>You buy a shield.%n");
                soldier.useCoins(2);
                soldier.addDefence();
                fightEnabled = false;
            }else{
                this.talk("You don't have enough coins.%n");
                fightEnabled = false;
            }
        }else if(choice.equalsIgnoreCase("3"))
        {
            System.out.print("=>You leave.%n");
            fightEnabled = false;
        }else{
            System.out.print("=>Illegal choice!%n");
        }
    }
    // If the soldier doesn't have enough coins to buy what (s)he wants, the merchant will say "You don't have enough coins.%n%n".
    // Then, the soldier will automatically leave.

    // After the soldier successfully buys an item from the merchant, (s)he will also automatically leave.
  }

  public void talk(String text) {
    System.out.printf("Merchant$: " + text);
  }

  // TODO: Other functions.
  public Pos getPos(){
      return this.pos;
  }
  
  public void setPos(int row, int column){
      this.pos.setPos(row, column);
  }
  
  public void displaySymbol(){
      System.out.print("$");
  }
       
}